### ISSUE TITLE
Multiple versions of the same package

### ISSUE THREAD

**Issue by sebdeckers** [myndzi/node-license-validator#5]
I'm running into an issue where my package has deep dependencies on multiple versions of the same package with different license string. I can't seem to permit more than one using `--allow-packages`.

Abbreviated example of the relevant data:

```
# Specify multiple versions of the same package as exceptions
> cat node-license-validator.json
{
  "deep": true,
  "production": true,
  "licenses": [ "MIT" ],
  "allow-packages": [
    "optimist@0.2.8",
    "optimist@0.6.1"
  ]
}

# Both versions exist within the dependency tree.
# Top level dependencies are nodemon@1.7.1 and standard@3.5.1 (to reproduce this problem)
> npm ls
...
├─┬ nodemon@1.7.1
...
│ ├─┬ ps-tree@0.0.3
│ │ └─┬ event-stream@0.5.3
│ │   └── optimist@0.2.8
...
├─┬ standard@5.3.1
...
│ ├─┬ standard-engine@2.2.0
...
│ │ ├─┬ eslint@1.6.0
│ │ │ ├─┬ handlebars@4.0.3
│ │ │ │ ├─┬ optimist@0.6.1
...

# Blows up! Note that v0.6.1 is exempted correctly.
> node-license-validator
Invalid license: optimist@0.2.8: MIT/X11
```

It looks like the problem lies in the mapping by package name which drops the version. Multiple versions will simply override each other.
See: https://github.com/myndzi/node-license-validator/blob/72df7bf274e0d526f81c88fc180b31a76feca617/index.js#L60

Any ideas on how to write a test case and fix for this? I looked but couldn't crack it just yet.

In the mean time my workaround is to have an exception for the package name without any version or use a semver range that includes both.

**Comment by myndzi** [comment]
Ah damn. I know better than this -- this is rooted in the original assumption that I wasn't bothering with deep matching.

Test case writing should be easy enough, just create an appropriately structured directory structure with a couple package.jsons ... fixing, I don't know. I'll have a look at it right now.

**Comment by myndzi** [comment]
...actually, you should be able to write a semver expression that allows explicitly these two versions. Try "0.2.8 || 0.6.1" as your allowed version

**Comment by myndzi** [comment]
(The config file should really be an object/map, but since yargs is managing it, it's an array. I could also concatenate items specified multiple times with `||`, or throw a more useful error such as "You specified the same package twice, is that a mistake?")

**Comment by sebdeckers** [comment]
That `||` did the trick, thanks! https://github.com/cofounders/lgtm/blob/master/node-license-validator.json#L24

As for warnings, I would also think it's useful to show when explicitly allowed licenses or excempted packages are _not_ encountered. That might indicate an out of date whitelist.

**Comment by myndzi** [comment]
Published 1.2.1 with warnings for both of these things; warnings are silenced with --quiet

**Comment by sebdeckers** [comment]
Awesome! :+1: BTW I recently gave a presentation about `node-license-validator` at the SingaporeJS meetup. Content is online if you're interested: https://github.com/cbas/talk-spdx

Strangely enough I'm seeing some false positives in the warning with v1.2.1.

```
Sebs-MacBook:lgtm seb$ node-license-validator
WARN: Packages were listed as exceptions but not found in the project: optimist, undefsafe, event-stream, ps-tree, rx-lite, sync-exec
Identified licenses: (BSD-2-Clause OR MIT OR Apache-2.0), (MIT AND CC-BY-3.0), (MIT OR Apache-2.0), Apache-2.0, BSD, BSD-2-Clause, BSD-3-Clause, CC-BY-3.0, CC-BY-4.0, ISC, MIT, Unlicense, WTFPL
All licenses ok.
Sebs-MacBook:lgtm seb$ npm ls | grep optimist
│ │ │   └─┬ optimist@0.2.8
│ │ │ │ ├─┬ optimist@0.6.1
Sebs-MacBook:lgtm seb$ npm ls | grep undefsafe
│ ├── undefsafe@0.0.3
Sebs-MacBook:lgtm seb$ npm ls | grep event-stream
│ │ └─┬ event-stream@3.3.2
│ │ │ └─┬ event-stream@0.5.3
Sebs-MacBook:lgtm seb$ npm ls | grep ps-tree
│ ├─┬ ps-tree@1.0.1
│ │ ├─┬ ps-tree@0.0.3
Sebs-MacBook:lgtm seb$ npm ls | grep rx-lite
│ │ │ │ ├── rx-lite@2.5.2
Sebs-MacBook:lgtm seb$ npm ls | grep sync-exec
│   │ │ └── sync-exec@0.5.0
```

The JSON config file is:

``` json
{
  "deep": true,
  "production": true,
  "allow-licenses": [
    "ISC",
    "MIT",
    "BSD",
    "WTFPL",
    "BSD-2-Clause",
    "BSD-3-Clause",
    "Apache-2.0",
    "CC-BY-3.0",
    "CC-BY-4.0",
    "Unlicense"
  ],
  "allow-packages": [
    "form-data@0.0.3",
    "install@0.1.8",
    "mime@1.2.7",
    "request@2.11.4",
    "wordwrap@0.0.2",
    "jsonify@0.0.0",
    "indexof@0.0.1",
    "optimist@0.2.8||0.6.1",
    "asn1@0.1.11",
    "aws-sign2@0.5.0",
    "commondir@0.0.1",
    "ctype@0.5.3",
    "htmlescape@1.1.0",
    "tunnel-agent@0.4.1",
    "undefsafe@0.0.3",
    "yargs@3.5.4",
    "event-stream@0.5.3",
    "mkdirp@0.3.0",
    "ps-tree@0.0.3",
    "rx-lite@2.5.2",
    "sync-exec@0.5.0",
    "flatten@0.0.1",
    "stream-spec@0.3.5",
    "stringify@3.1.0"
  ]
}
```

The repo is https://github.com/cofounders/lgtm
